airbyte-integrations/connectors/destination-snowflake/src/main/kotlin/io/airbyte/integrations/destination/snowflake/typing_deduping/SnowflakeSqlGenerator.kt

```diff
@@ -211,47 +320,124 @@ class SnowflakeSqlGenerator(private val retentionPeriodDays: Int) : SqlGenerator
         minRawTimestamp: Optional<Instant>,
         useTryCast: Boolean
     ): String {
-        val columnList = stream.columns.keys.joinToString("\n") { "${it.name(QUOTE)}," }
-        val extractNewRawRecords = extractNewRawRecords(stream, minRawTimestamp, useTryCast)
+        val selectRecordsFromRawTable =
+            selectTypedRecordsFromRawTable(stream, minRawTimestamp, useTryCast)
 
         return """
-            |INSERT INTO ${stream.id.finalTableId(QUOTE, finalSuffix.uppercase(Locale.getDefault()))}(
-            |  $columnList
-            |  "_AIRBYTE_META",
-            |  "_AIRBYTE_RAW_ID",
-            |  "_AIRBYTE_EXTRACTED_AT",
-            |  "_AIRBYTE_GENERATION_ID"
+            |INSERT INTO ${stream.id.quotedFinalTableId(finalSuffix)}(
+            |${stream.finalTableColumnsWithoutMeta { it }.replaceIndent("  ")},
+            |  "_AIRBYTE_META"
             |)
-            |$extractNewRawRecords;
+            |$selectRecordsFromRawTable;
             |""".trimMargin()
     }
 
-    private fun extractNewRawRecords(
+    private fun StreamConfig.projectionColumnsFromRawTable(safeCast: Boolean): String {
+        val adjustedExtractedAt =
+            "${airbyteExtractedAtUtcForced(JavaBaseConstants.COLUMN_NAME_AB_EXTRACTED_AT.quoted())} as " +
+                JavaBaseConstants.COLUMN_NAME_AB_EXTRACTED_AT.quoted()
+
+        // All columns including _ab_. This safeguards against empty data columns
+        // and phantom commas in sql because we always have meta columns.
+        val columns =
+            this.columns.map {
+                "${
+                extractAndCast(
+                    it.key,
+                    it.value,
+                    safeCast,
+                )
+            } as ${it.key.name.quoted()}"
+            } +
+                listOf(
+                    JavaBaseConstants.COLUMN_NAME_AB_RAW_ID.quoted(),
+                    adjustedExtractedAt,
+                    JavaBaseConstants.COLUMN_NAME_AB_META.quoted(),
+                    JavaBaseConstants.COLUMN_NAME_AB_GENERATION_ID.quoted(),
+                )
+
+        return columns.joinToString(", \n")
+    }
+
+    private fun StreamConfig.typeCastError(safeCast: Boolean): String {
+        val caseStatement: (Map.Entry<ColumnId, AirbyteType>) -> String = { col ->
+            """
+                |CASE
+                |  WHEN (TYPEOF("_airbyte_data":"${escapeJsonIdentifier(col.key.originalName)}") NOT IN ('NULL', 'NULL_VALUE'))
+                |    AND (${extractAndCast(col.key, col.value, safeCast)} IS NULL)
+                |    THEN OBJECT_CONSTRUCT('field', '${escapeSingleQuotedString(col.key.originalName)}', 'change', 'NULLED', 'reason', 'DESTINATION_TYPECAST_ERROR')
+                |  ELSE NULL
+                |END
+            """.trimMargin()
+        }
+        return this.columns.map(caseStatement).joinToString(", \n")
+    }
+
+    private fun StreamConfig.finalTableColumnsWithoutMeta(
```

Comment:
why does meta need to be filtered out here?

Reply:
it has 3 different variations which aren't the same (like constructing meta, just selecting meta etc) so ended up leaving it out from the common method.